### ISSUE TITLE
NullPointerException when aggregating row value expressions using ARRAY_AGG()

### ISSUE THREAD

**Issue by JanecekPetr** [jOOQ/jOOQ#7691]
### Expected behavior and actual behavior:
Imagine a table like this:

```sql
CREATE TABLE bang (
    foo INTEGER NOT NULL,
    bar INTEGER NOT NULL
);
```

and a jooq select like this:

```java
private static final Field<Record2<Integer, Integer>[]> BANG_AGGREGATION = arrayAgg(
        rowField(row(BANG.FOO, BANG.BAR))
    ).as("foobar");

...

context.select(BANG_AGGREGATION)
    .from(BANG)
    .fetchSingle(BANG_AGGREGATION);
```

It kinda fails. It throws an exception in the middle of execution, but survives it and returns a record with nulls instead of the values received from the DB. I expected it to not fail.

The rendered SQL is good:

```sql
select array_agg((bang.foo, bang.bar)) as foobar
from bang
```

The DB even responses with the expected values, but then:

```
18:16:59.323 [main] DEBUG org.jooq.tools.LoggerListener - Executing query          : select array_agg(("public"."bang"."foo", "public"."bang"."bar")) as "foobar" from "public"."bang"
18:16:59.368 [main] ERROR org.jooq.impl.DefaultBinding - Cannot parse array       : {"(1,2)","(1,2)"}
java.lang.IllegalStateException: Could not construct new record
	at org.jooq.impl.Tools.newRecord(Tools.java:765) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.Tools.newRecord(Tools.java:748) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.Tools.newRecord(Tools.java:712) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.pgNewRecord(DefaultBinding.java:3017) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:2857) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:2778) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:774) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.pgGetArray(DefaultBinding.java:1095) [jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:1032) [jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$DefaultArrayBinding.get0(DefaultBinding.java:911) [jooq-3.11.3.jar:?]
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:774) [jooq-3.11.3.jar:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1720) [jooq-3.11.3.jar:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1689) [jooq-3.11.3.jar:?]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1654) [jooq-3.11.3.jar:?]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:125) [jooq-3.11.3.jar:?]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1618) [jooq-3.11.3.jar:?]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1585) [jooq-3.11.3.jar:?]
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:407) [jooq-3.11.3.jar:?]
	at org.jooq.impl.Tools.fetchSingle(Tools.java:1814) [jooq-3.11.3.jar:?]
	at org.jooq.impl.AbstractResultQuery.fetchSingle(AbstractResultQuery.java:640) [jooq-3.11.3.jar:?]
	at org.jooq.impl.AbstractResultQuery.fetchSingle(AbstractResultQuery.java:580) [jooq-3.11.3.jar:?]
	at org.jooq.impl.SelectImpl.fetchSingle(SelectImpl.java:2909) [jooq-3.11.3.jar:?]
	at com.gitlab.janecekpetr.jooqboom.Kabang.selectAggregatedRowValueExpressions(Kabang.java:58) [classes/:?]
	at com.gitlab.janecekpetr.jooqboom.Kabang.selectPlease(Kabang.java:38) [classes/:?]
	at com.gitlab.janecekpetr.jooqboom.KabangTest.testSelectAggregatedRowValueExpressions(KabangTest.java:15) [test-classes/:?]
    ...
Caused by: java.lang.NullPointerException
	at org.jooq.impl.Fields.size(Fields.java:81) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.RowImpl.size(RowImpl.java:224) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.AbstractRecord.<init>(AbstractRecord.java:136) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.RecordImpl.<init>(RecordImpl.java:131) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.Tools$1.newInstance(Tools.java:782) ~[jooq-3.11.3.jar:?]
	at org.jooq.impl.Tools.newRecord(Tools.java:756) ~[jooq-3.11.3.jar:?]
	... 47 more
```

I'm aware that `rowField` is marked EXPERIMENTAL, and apparently rightly so. Even then, one can always hope... :)

### Steps to reproduce the problem (include example code if possible):
If you want, https://gitlab.com/janecekpetr/jooq-boom contains the problem under the name Kabang. But the code above should be enough to reproduce easily.

### Versions:

- jOOQ: 3.11.3
- Java: 8u141
- Database (include vendor): PostgreSQL 9.5.12

- OS: Windows 7
- JDBC Driver (include name if inofficial driver): pgjdbc 42.2.4

**Comment by lukaseder** [comment]
Thank you very much for your report. That's indeed something quite desirable, specifically in the case of `array_agg()`. Most other aggregate functions regrettably do not support row value expressions, e.g. `min()` and `max()` don't, even if the behaviour would be straightforward. So, while your API usage should be fixed, I could also see adding `arrayAgg(Row[N]<...>)` methods

**Comment by lukaseder** [comment]
Related issue that has produced an incomplete fix: #6404

**Comment by lukaseder** [comment]
This is now fixed for jOOQ 3.12. Backport scheduled for 3.11.5 (#7784).

Note that this, too, is an incomplete fix. Another problem remains: #7783. I.e. the fact that the data type meta model is not sophisticated enough to maintain the necessary information for nested record types of several levels, possibly leading to class cast exceptions.

**Comment by lukaseder** [comment]
Having said so, `DSL.rowField()` ist still rather experimental, as disclaimed in the Javadoc :) Thanks for trying it out and reporting issues!
